vminitd/Sources/vminitd/ManagedContainer.swift

```diff
@@ -28,6 +36,370 @@ actor ManagedContainer {
     private let log: Logger
     private let bundle: ContainerizationOCI.Bundle
     private var execs: [String: ManagedProcess] = [:]
+    private var namespaceWorker: NamespaceWorker?
+
+    /// Worker child process that runs in container's namespace for filesystem operations
+    private final class NamespaceWorker: @unchecked Sendable {
+        private let containerID: String
+        private let containerPID: Int32
+        private var childPID: Int32?
+        private var parentSocket: Int32?
+        private var eventIDCounter: UInt32 = 0
+        private let pendingEvents: Mutex<[UInt32: CheckedContinuation<Void, Error>]> = Mutex([:])
+        private var responseReaderTask: Task<Void, Never>?
+        private let shouldStop: Atomic<Bool> = Atomic(false)
+
+        init(containerID: String, containerPID: Int32) {
+            self.containerID = containerID
+            self.containerPID = containerPID
+        }
+
+        func start() throws {
+            guard childPID == nil else {
+                throw ContainerizationError(.invalidState, message: "NamespaceWorker already started")
+            }
+
+            // Create socketpair for parent-child communication
+            var sockets: [Int32] = [0, 0]
+            guard socketpair(AF_UNIX, SOCK_STREAM, 0, &sockets) == 0 else {
+                throw ContainerizationError(.internalError, message: "Failed to create socketpair: errno \(errno)")
+            }
+
+            let parentSocket = sockets[0]
+            let childSocket = sockets[1]
+
+            // Fork child process
+            let pid = fork()
+            guard pid >= 0 else {
+                close(parentSocket)
+                close(childSocket)
+                throw ContainerizationError(.internalError, message: "Failed to fork: errno \(errno)")
+            }
+
+            if pid == 0 {
+                // Child process
+                close(parentSocket)
+                runChildProcess(socket: childSocket)
+                exit(0)
+            } else {
+                // Parent process
+                close(childSocket)
+                self.childPID = pid
+                self.parentSocket = parentSocket
+
+                // Wait for child to signal ready or failure
+                var signal: UInt8 = 0
+                let readResult = read(parentSocket, &signal, 1)
+
+                if readResult != 1 {
+                    // Child failed to send signal
+                    close(parentSocket)
+                    self.parentSocket = nil
+                    var status: Int32 = 0
+                    waitpid(pid, &status, 0)
+                    self.childPID = nil
+                    throw ContainerizationError(.internalError, message: "Child process failed to start")
+                }
+
+                if signal == 0xFF {
```

Comment:
Extract constants for these and then you can get rid of the comments.

Also, let's not call it a "signal" as that has a specific meaning that is not this.